src/index.js

```diff
@@ -153,15 +152,48 @@ module.exports = class Bitwap {
   }
 
   getBlocks (keys, cb) {
-    const blocks = []
-    const finish = (block) => {
-      blocks.push(block)
-      log('finish: %s/%s', blocks.length, keys.length)
-      if (blocks.length === keys.length) {
-        cb(null, blocks)
+    const results = {}
+    const unwantListeners = {}
+    const blockListeners = {}
+    const unwantEvent = (key) => `unwant:${key.toString('hex')}`
+    const blockEvent = (key) => `block:${key.toString('hex')}`
+
+    const cleanupListeners = () => {
+      keys.forEach((key) => {
+        this.notifications.removeListener(unwantEvent(key), unwantListeners[key])
+        this.notifications.removeListener(blockEvent(key), blockListeners[key])
+      })
+    }
+
+    const addListeners = () => {
+      keys.forEach((key) => {
+        unwantListeners[key] = () => {
+          finish(key, new Error(`manual unwant: ${key.toString('hex')}`))
+        }
+
+        blockListeners[key] = (block) => {
+          finish(key, null, block)
+        }
+
+        this.notifications.once(unwantEvent(key), unwantListeners[key])
+        this.notifications.once(blockEvent(key), blockListeners[key])
+      })
+    }
+
+    const finish = (key, err, block) => {
+      results[key] = {
+        error: err,
+        block: block
+      }
+
+      if (Object.keys(results).length === keys.length) {
+        cleanupListeners()
+        cb(results)
       }
     }
 
+    addListeners()
+
     keys.forEach((key) => {
       // Sanity check, we don't want to announce looking for blocks
       // when we might have them ourselves
```

Comment:
This is more than sanity, since now the Block Service has an 'online' mode that defers everything to bitswap